onnxruntime/core/optimizer/embed_layer_norm_fusion.cc

```diff
@@ -184,60 +184,68 @@ static bool MatchPositionSubgraph(
-*/
-static bool MatchPositionEmbeddingSubgraph1(
+ * 
+ *    Shape -> ^Gather (indice=0)^ -> ^Unsqueeze^
+ *      /                                  |           +-----------------------+
+ *     /                                   v           |                       |
+ * [input_ids]                          ^Concat^ -> *Reshape* -> *Equal* -> *Where* -> Expand -> Gather
+ *     \                                   |                                            |   ("position")
+ *    Shape -> ^Gather (indice=1)^ -> ^Unsqueeze^                                       |
+ *                |                                                                     |
+ *                +-------------- # one of the below subgraph patterns # ---------------+
+ *       # Unsqueeze -> ConstantOfShape -> NonZero -> Transpose -> Squeeze -> Cast -> Unsqueeze #
+ *       #                                      or                                              #
+ *       #        Cast (to=7) (optional) -> Range (start=0, delta=1) -> Unsqueeze               #
+ * 
+ * Note that position gather node is the node in the bottom of above sub-graph.
+ * Paths in ^^ are alternative path to be matched if path input_ids -> Shape -> Expand -> Gather is not found.
+ * Path in ** is an alternative path to check.
+ */
+static bool MatchPositionEmbeddingSubgraphsFromGather(
     Graph& graph,
     const Node& position_gather_node,
     const NodeArg* input_ids,
     const logging::Logger& logger,
     std::vector<NodeIndex>& subgraph_node_indices) {
   subgraph_node_indices.clear();
-
   std::vector<const Node::EdgeEnd*> pg_edges;
   // Look for Path 1:
-  // Shape --> Gather --> Unsqueeze --> ConstantOfShape --> NonZero --> Transpose --> Squeeze --> Cast --> Unsqueeze --> Expand --> Gather
-  if (!graph_utils::FindPath(position_gather_node, true,
-                             {{0, 1, "Expand", {8}, kOnnxDomain},
-                              {0, 0, "Unsqueeze", {1, 11}, kOnnxDomain},
-                              {0, 0, "Cast", {9}, kOnnxDomain},
-                              {0, 0, "Squeeze", {1, 11}, kOnnxDomain},
-                              {0, 0, "Transpose", {1}, kOnnxDomain},
-                              {0, 0, "NonZero", {9}, kOnnxDomain},
-                              {0, 0, "ConstantOfShape", {9}, kOnnxDomain},
-                              {0, 0, "Unsqueeze", {1, 11}, kOnnxDomain},
-                              {0, 0, "Gather", {1, 11}, kOnnxDomain},
-                              {0, 0, "Shape", {1}, kOnnxDomain}},
-                             pg_edges, logger)) {
+  // Shape --> Gather --> Unsqueeze --> ConstantOfShape --> NonZero --> Transpose --> Squeeze
+  // --> Cast --> Unsqueeze --> Expand --> Gather
+  std::vector<graph_utils::EdgeEndToMatch> parent_path_1{
+      {0, 1, "Expand", {8}, kOnnxDomain},
+      {0, 0, "Unsqueeze", {1, 11}, kOnnxDomain},
+      {0, 0, "Cast", {9}, kOnnxDomain},
```

Comment:
Cast might be removed by CastElimination optimizer if input and output has same data type.